core/workspace_svg.js

```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
   Blockly.mainWorkspace = this;
 };
 
+/**
+ * Zooming the blocks centered in (x, y) coordinate with zooming in or out.
+ * @param {!number} x X coordinate of center.
+ * @param {!number} y Y coordinate of center.
+ * @param {!number} type Type of zomming (-1 zooming out and 1 zooming in).
+ */
+Blockly.WorkspaceSvg.prototype.zoom  = function(x, y, type) {
+  var speed = this.scaleSpeed;
+  var metrics = this.getMetrics();
+  var center = this.options.svg.createSVGPoint();
+  var g = this.getCanvas();
+  center.x = x;
+  center.y = y;
+  center = center.matrixTransform(workspace.getCanvas().getCTM().inverse());
+  var x = center.x;
+  var y = center.y;
+  var canvas = workspace.getCanvas();
+  // scale factor
+  var scale = (type == 1)?speed:1/speed;
+  var matrix = canvas.getCTM().translate(-(x*(scale-1)),-(y*(scale-1))).scale(scale);
+  // validate if scale is in a valid range
+  if (matrix.a >= this.minScale && matrix.a <= this.maxScale) {
+    this.scale = matrix.a;
+    this.scrollX = matrix.e - metrics.absoluteLeft;
+    this.scrollY = matrix.f - metrics.absoluteTop ;
+    this.updateGridPattern_();
+    this.scrollbar.resize();
+  }
+}
+
+/**
+ * Zooming the blocks centered in the center of view with zooming in or out.
+ * @param {!number} type Type of zomming (-1 zooming out and 1 zooming in).
+ */
+Blockly.WorkspaceSvg.prototype.zoomCenter  = function(type) {
+  var metrics = this.getMetrics();
+  var x = metrics.viewWidth/2;
+  var y = metrics.viewHeight/2;
+  this.zoom(x ,y , type);
+}
+
+/**
+ * Reset zooming and dragging.
+ */
+Blockly.WorkspaceSvg.prototype.zoomReset  = function() {
+  var metrics = this.getMetrics();
+  Blockly.hideChaff();
+  workspace.scrollbar.set(-metrics.contentLeft, -metrics.contentTop);
+  this.scale = 1;
+  this.updateZoom();
+  this.updateGridPattern_();
+}
+
+/**
+ * Updates the grid pattern.
+ */
+Blockly.WorkspaceSvg.prototype.updateGridPattern_  = function() {
+  this.options.gridPattern.setAttribute('width', this.options.gridOptions.spacing * this.scale+'');
```

Comment:
Setting an attribute does not need string conversion.  The browser will handle it.